lib/srv/server/cloudwatcher.go

```diff
@@ -0,0 +1,197 @@
+
+	fetchers      []*ec2InstanceFetcher
+	fetchInterval time.Duration
+	ctx           context.Context
+	cancel        context.CancelFunc
+}
+
+func (w *Watcher) Run() {
+	ticker := time.NewTicker(w.fetchInterval)
+	defer ticker.Stop()
+	for {
+		for _, fetcher := range w.fetchers {
+			inst, err := fetcher.GetEC2Instances(w.ctx)
+			if err != nil {
+				log.WithError(err).Error("Failed to fetch EC2 instances")
+				continue
+			}
+			if inst == nil {
+				continue
+			}
+			select {
+			case w.InstancesC <- *inst:
+			case <-w.ctx.Done():
+			}
+		}
+		select {
+		case <-ticker.C:
+			continue
+		case <-w.ctx.Done():
+			return
+		}
+	}
+}
+
+func (w *Watcher) Stop() {
+	w.cancel()
+}
+
+func NewCloudServerWatcher(ctx context.Context, matchers []services.AWSMatcher, clients cloud.Clients) (*Watcher, error) {
+	cancelCtx, cancelFn := context.WithCancel(ctx)
+	watcher := Watcher{
+		fetchers:      []*ec2InstanceFetcher{},
+		ctx:           cancelCtx,
+		cancel:        cancelFn,
+		fetchInterval: time.Minute,
+		InstancesC:    make(chan EC2Instances),
+	}
+	for _, matcher := range matchers {
+		for _, region := range matcher.Regions {
+			cl, err := clients.GetAWSEC2Client(region)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			fetcher := newEC2InstanceFetcher(ec2FetcherConfig{
+				Matcher:   matcher,
+				Region:    region,
+				Document:  matcher.SSM.Document,
+				EC2Client: cl,
+				Labels:    matcher.Tags,
+			})
+			watcher.fetchers = append(watcher.fetchers, fetcher)
+		}
+	}
+	return &watcher, nil
+}
+
+type ec2FetcherConfig struct {
+	Matcher   services.AWSMatcher
+	Region    string
+	Document  string
+	EC2Client ec2iface.EC2API
+	Labels    types.Labels
+	VPC       string
+}
+
+type ec2InstanceFetcher struct {
+	Filters    []*ec2.Filter
+	EC2        ec2iface.EC2API
+	Region     string
+	Document   string
+	Parameters map[string]string
+}
+
+func newEC2InstanceFetcher(cfg ec2FetcherConfig) *ec2InstanceFetcher {
+	tagFilters := []*ec2.Filter{&ec2.Filter{
+		Name:   aws.String(constants.AWSInstanceStateName),
+		Values: aws.StringSlice([]string{ec2.InstanceStateNameRunning}),
+	}}
+	if cfg.VPC != "" {
+		tagFilters = append(tagFilters, &ec2.Filter{
+			Name:   aws.String(constants.AWSInstanceStateName),
+			Values: aws.StringSlice([]string{cfg.VPC}),
+		})
+	}
+
+	for key, val := range cfg.Labels {
+		tagFilters = append(tagFilters, &ec2.Filter{
+			Name:   aws.String("tag:" + key),
+			Values: aws.StringSlice(val),
+		})
+	}
+	fetcherConfig := ec2InstanceFetcher{
+		EC2:      cfg.EC2Client,
+		Filters:  tagFilters,
+		Region:   cfg.Region,
+		Document: cfg.Document,
+		Parameters: map[string]string{
+			"token": cfg.Matcher.Params.JoinToken,
+		},
+	}
+	return &fetcherConfig
+}
+
+func (f *ec2InstanceFetcher) GetEC2Instances(ctx context.Context) (*EC2Instances, error) {
+	var instances []*ec2.Instance
+	var accountID string
+	err := f.EC2.DescribeInstancesPagesWithContext(ctx, &ec2.DescribeInstancesInput{
```

Comment:
This fetches all EC2 nodes matching the filters on every "tick" of the watcher, correct? I wonder if it's going to cause issues with large fleets of nodes.

We can probably address this out of scope of this particular PR, but I wonder if we can mark every AWS instance that we've already enrolled with some tag (e.g. `TeleportAutoEnrolled: true`) and then filter them out during this discovery. What do you think?